libs/ngxtension/lazy-service/src/lazy-service.ts

```diff
@@ -0,0 +1,39 @@
+import { Injector, Type, runInInjectionContext } from '@angular/core';
+import { assertInjector } from 'ngxtension/assert-injector';
+import { Observable, defer } from 'rxjs';
+
+/**
+ * Loads a service lazily. The service is loaded when the observable is subscribed to.
+ *
+ * @param loader A function that returns a promise of the service to load.
+ * @param injector The injector to use to load the service. If not provided, the current injector is used.
+ * @returns An observable of the service.
+ *
+ * @example
+ * ```ts
+ * const dataService$ = lazyService(() => import('./data-service').then((m) => m.MyService));
+ * or
+ * const dataService$ = lazyService(() => import('./data-service'));
+ * ```
+ */
+export function lazyService<T>(
+	loader: () => Promise<Type<T>> | Promise<{ default: Type<T> }>,
+	injector?: Injector
+): Observable<T> {
+	injector = assertInjector(lazyService, injector);
+
+	return runInInjectionContext(injector, () => {
+		return defer(() => {
+			return loader()
+				.then((serviceOrDefault) => {
+					if ('default' in serviceOrDefault) {
+						return injector!.get(serviceOrDefault.default);
+					}
+					return injector!.get(serviceOrDefault);
```

Comment:
If the service uses `DestroyRef.onDestroy()` it will never be called. 

Even if `injector` is a `NodeInjector`, this works only with `providedIn: root`. So it's the root injector  that will provide the `DestroyRef` (and thus never call `OnDestroy`. 

The solution would be to create an `EnvironmentInjector` that provides the class we just lazy-loaded. 

Something like : 
```
    if (!(injector instanceof EnvironmentInjector)) {
      // this is the DestroyRef of the component
      const destroyRef = injector.get(DestroyRef);

      // This is the parent injector of the injector we're creating
      const environmentInjector = injector.get(EnvironmentInjector);

      // Creating an environment injector to destroy it afterwards
      const newInjector = createEnvironmentInjector([type as Provider], environmentInjector);

      // Destroy the injector to trigger DestroyRef.onDestroy on our service
      destroyRef.onDestroy(() => {
        newInjector.destroy();
      });

      // We want to create the new instance of our service with our new injector
      injector = newInjector;
    }
```